### DISCUSSION TITLE
Can't connect to MySQL on a fresh install

### DISCUSSION THREAD

**Discussion by joeldcanfield** [wintercms/winter#157] [Q&A] [answered]
Pretty comfortable with the command line but I'm missing something about connecting Winter to MySQL on my local Mac (I have a photos app called Piwigo working fine, so the database and local MAMP server seem to be working.)

After I run `php artisan winter:install` I go through all the info and it returns this big red error:

`SQLSTATE[HY000] [2002] No such file or directory (SQL: select * from information_schema.tables where table_schema = winter and table_name = migrations and table_type = 'BASE TABLE')`

It appears to be looking for an existing database, but there's nothing in the command line install procedure about creating and/or configuring said database.

ALSO

when I go to /mywinter/ in the browser, I get THIS big red error:

`An exception has been thrown during the compilation of a template ("SQLSTATE[HY000] [1045] Access denied for user 'spinhead'@'localhost' (using password: YES) (SQL: select * from `system_parameters` where `namespace` = system and `group` = core and `item` = build limit 1)").`

NOTE
I feel like these two things are very closely related, and that I should see it. But my brain is oatmeal today, which is why I thought it was a good day to discover and install and brand new cms I'd never heard of. Be kind to my oatmeal.

Thanks.

joel

**Comment by bennothommo** [comment, marked as answer]
@joeldcanfield It looks like your database credentials in the config are incorrect. You'll need to double-check those in `config/database.php`.

Final solution for this specific issue: Needed to use `127.0.0.1` instead of `localhost` for the database host option.

**Reply by joeldcanfield** [reply]
Found some control characters at the end of the password. Deleted those (thanks for the pointer) and now I'm getting a different error in the page, but same error in the winter:install command line.

`SQLSTATE[42S02]: Base table or view not found: 1146 Table 'winter.system_settings' doesn't exist (SQL: select * from `system_settings` where `item` = system_log_settings limit 1)`

**Comment by mjauvin** [comment]
@joeldcanfield did you run "artisan winter:up" after creating the database?

**Reply by joeldcanfield** [reply]
No. In case it is unclear, this is a bare naked brand new install. Not migrating anything. Running `winter:up` gives this error

`SQLSTATE[HY000] [2002] No such file or directory (SQL: select * from information_  
  schema.tables where table_schema = winter and table_name = migrations and table_t  
  ype = 'BASE TABLE') `

followed by others of the same ilk.

**Reply by bennothommo** [reply]
@joeldcanfield We still recommend using `php artisan winter:up` with a fresh install too - it will initialise a fresh database. `php artisan winter:install` is used to configure your database and site settings before it installs the database, and it seems from your comments that it perhaps is adding weird characters to the config.

The steps you should take are:
- Install Winter CMS into a directory.
- Set up configuration in the `config` directory (you can also use `php artisan winter:env` to do configuration through a `.env` file and use environment variables)
- Run `php artisan winter:up`

Give that a try and see how you go with that.

**Reply by joeldcanfield** [reply]
I have gone through the process again, and at every turn I get an error saying there is no table named 'migrations' (which I can confirm, because the database is entirely empty.)

**Reply by bennothommo** [reply]
Just to try something else, can you create another empty database, update your config to use this new database, and try the commands again?

**Reply by joeldcanfield** [reply]
Good thought, but no joy.

Is winter:install supposed to create the table 'migrations' which the error keeps referring to? When does that table get created, since it's the error that keeps recurring?

**Reply by bennothommo** [reply]
@joeldcanfield The Laravel framework itself that Winter CMS is built on is supposed to set up the `migrations` table. It does this if the database is empty. With the steps above that I posted, did you run the `winter:up` command only? Or have you run `winter:install`?

**Reply by joeldcanfield** [reply]
_From above:_

**After I run php artisan winter:install I go through all the info and it returns this big red error:**

`SQLSTATE[HY000] [2002] No such file or directory (SQL: select * from information_schema.tables where table_schema = winter and table_name = migrations and table_type = 'BASE TABLE')`

My first error was that when running install, it couldn't find the table. Which confuses me (still, or again, or more) since you say that step is supposed to create the db.

Later today (haven't had my morning tea yet) I'm going to nuke the whole thing and start again. It seems something's bugged up at an earlier/fundamental point. Though if anyone has any simpler tips (not that going nuclear will be that difficult) I'm all ears. I do appreciate the support here and hope I haven't sounded shirty.

**Reply by bennothommo** [reply]
We will have a web based installer very shortly which might have better luck for you in getting started, if the full nuke and refresh doesn't work.

**Reply by joeldcanfield** [reply]
During the fresh install I got this error more than once. It probably happened the first time and I didn't pay attention.

`*** No database detected - skipping setting the build number.`

If that's the issue, how do I ensure it can see the db?

**Reply by LukeTowers** [reply]
That will happen the first time it gets run before the database exists yet

**Reply by joeldcanfield** [reply]
Okay, but then when I get to winter:install I still get the table-not-found error.

I really want this to work, only because I'm curious about the Winter CMS project. Truly prefer not to wait for a web installer, which I'd prefer not to use anyway.

Do others have this much trouble with the install?

**Reply by LukeTowers** [reply]
No, so far it's only been you. Did you try using SQLite instead?

**Reply by joeldcanfield** [reply]
I'd rather not. I've spent years in MySQL and my curiosity only extends so far.

So, clarifying again: when, precisely, in the setup and config, is the migration table created in the database?

**Reply by bennothommo** [reply]
@joeldcanfield 

> So, clarifying again: when, precisely, in the setup and config, is the migration table created in the database?

I mentioned this in my previous post: The Laravel framework itself that Winter CMS is built on is supposed to set up the migrations table. It does this if the database is empty.  I'll add that this happens when running migrations for the first time.

You also seem to keep mentioning the `winter:install` command - can you please confirm you tried my previous steps exactly with a fresh copy and database? If you follow my steps, then there's no need to use the `winter:install` command at all.

**Reply by joeldcanfield** [reply]
in a folder called root I ran this command

`composer create-project wintercms/winter mywinter`

I looked through cms.php and database.php in /mywinter/config/ and found nothing that needed changing except as instructed in the cli install instructions

`'disableCoreUpdates' => true,`

Ran php artisan winter:up from /mywinter/ and it returned this

`Migrating application and plugins...

In Connection.php line 669:                                                             
  SQLSTATE[HY000] [2002] No such file or directory (SQL: select * from information_schema.tables where table_sc  
  hema = database and table_name = migrations and table_type = 'BASE TABLE')                                     
                                                                                                                 
In Exception.php line 18:
  SQLSTATE[HY000] [2002] No such file or directory  
                                                    
In PDOConnection.php line 39:
  SQLSTATE[HY000] [2002] No such file or directory `

**Reply by bennothommo** [reply]
@joeldcanfield as @mjauvin has mentioned below, you will need to add your database credentials to `config/database.php` *before* running `winter:up`, in the following section:

```php
        'mysql' => [
            'driver'     => 'mysql',
            'engine'     => 'InnoDB',
            'host'       => 'localhost',
            'port'       => 3306,
            'database'   => 'database',
            'username'   => 'root',
            'password'   => '',
            'charset'    => 'utf8mb4',
            'collation'  => 'utf8mb4_unicode_ci',
            'prefix'     => '',
            'varcharmax' => 191,
        ],
```

Unless these truly are the settings you use for the database, in which case I would strongly recommend securing your MAMP setup.

**Reply by joeldcanfield** [reply]
@mjauvin @bennothommo Sigh. Thanks. Appreciate the patience. No, those aren't my db settings ;) Off to wipe the slate and walk s l o w l y through the process, using what some might call "the right way."

**Reply by LukeTowers** [reply]
@joeldcanfield it's worth noting that the `winter:install` command can be used (if you aren't already using env() calls in your config files) to change those values for you as you go through the installation questionnaire, but you do still have to provide the correct details to that command in order for the config files to be appropriately changed.

My own personal preference is always to manually edit the files to what they need to be and then run `winter:up`, but `winter:install` is meant to simplify that process somewhat for people.

**Reply by joeldcanfield** [reply]
Here is the exact process I followed, with my eagle-eyed wife watching to catch typos and stupidity on my part:

1. in terminal on my Mac I went to `~/root/`
2. ran the command `composer create-project wintercms/winter mywinter`
3. edited `/mywinter/config/database.php` to add the database name, username, and password to the mysql settings section
4. moved back to `/mywinter/`
5. ran `php artisan winter:up`
6. it returned this


`Migrating application and plugins...

In Connection.php line 669:                                                                                    
  SQLSTATE[HY000] [2002] No such file or directory (SQL: select * from information_schema.tables where table_sc  
  hema = winterten and table_name = migrations and table_type = 'BASE TABLE')                                    

In Exception.php line 18:                        
  SQLSTATE[HY000] [2002] No such file or directory  

In PDOConnection.php line 39:                   
  SQLSTATE[HY000] [2002] No such file or directory  `

**Reply by LukeTowers** [reply]
@joeldcanfield what is the output of `php --version`, and what is the configuration you're using for the PHP project that can connect to MySQL?

**Reply by joeldcanfield** [reply]
php --version
PHP 7.3.28 (cli) (built: Apr 29 2021 14:53:09) ( NTS )

The photos tool connects using the same username and password I'm trying here, or at least, that's how it appears at the moment. I'm going to fiddle with it over the weekend and make sure I'm not doing something dumb with the db connection.

**Reply by bennothommo** [reply]
@joeldcanfield The error `SQLSTATE[HY000] [2002] No such file or directory` means that it cannot establish a connection with the database at all - not even to authenticate.

You could perhaps have a look at the following Stack Overflow question which has a few different fixes to try:
https://stackoverflow.com/questions/20723803/pdoexception-sqlstatehy000-2002-no-such-file-or-directory

**Reply by joeldcanfield** [reply]
Success. In config/database.php I changed host from `localhost` to `127.0.0.1` (and here I thought learning about the ISO/OSI all those years ago wouldn't come in handy) and boom.

Thanks for the ongoing support. A couple thoughts:

1. I mentioned the specific error `SQLSTATE[HY000] [2002] No such file or directory` in my original post. Having just now googled that phrase, the solution is the first hit. Curious that it took 9 days to come around to that being the bit that mattered.
2. I cannot see how anyone would arrive at this precise installation procedure from the instructions at the website. Are you guys sure people are installing this without trouble? I can't imagine I'm the only MAMP user (that seems to be the core of it) who's trying to install Winter and running into this blind loop.

Lest I sound like an angry whiner, I am, in fact, a huge fan of what you're doing. Knowing that you've launched Winter because October is no longer open source, reading your reasons etc., mattered to me. If I hadn't been really persistent, I'd have left with a negative impression of the whole project, which is what I don't want happening to other people..

Now I sound like it's the third round at the pub instead of 9am. Anyway, thanks for seeing this through @bennothommo @LukeTowers @mjauvin

**Reply by LukeTowers** [reply]
> Curious that it took 9 days to come around to that being the bit that mattered.

Yup, hence why I usually recommend people google their errors, especially errors that come from other pieces of the stack, first, lots of times it's got nothing to do with Winter.

**Reply by joeldcanfield** [reply]
Always. But since I googled the entire error and not just the initial phrase, everything I found pertained to the `(SQL: select * from information_schema.tables where table_schema = winter and table_name = migrations and table_type = 'BASE TABLE')` and pointed me back here. Since @bennothommo explicitly says in his comment _The error **SQLSTATE[HY000] [2002] No such file or directory** means that it cannot establish a connection with the database at all - not even to authenticate._ and since that is exactly what led to discovering the solution, well, I'm not sure what I planned to point out, but I don't want to get into a fuss over it.

Not having to do with Winter is, perhaps, not entirely accurate. Every tool I use to connect to my local instance of MySQL has been fine with `localhost`. (WordPress, Piwigo, others.) Only Winter has failed in a way that required changing `localhost` to the IP address. It may well be connected to MAMP, and be a very specific conflict, but it's one worth knowing about.

**Reply by LukeTowers** [reply]
The problem is most likely that MAMP has an incorrect path to the mysql socket file in its PHP configuration, and on unix systems `localhost` attempts to use sockets instead of TCP.

**Reply by joeldcanfield** [reply]
I wondered about that, but in php.ini

`pdo_mysql.default_socket = /Applications/MAMP/tmp/mysql/mysql.sock`

and in the terminal

`[joel@Galway.local /Applications/mamp/tmp/mysql]$ ls
my.cnf          mysql.pid       **mysql.sock**      mysql.sock.lock tmpdir`

so I'm not sure what's up. However, MAMP does appear to have some challenge with using sockets for MySQL, if the number of Google results is any indication.

**Reply by mjauvin** [reply]
please note that filenames are CASE SENSITIVE on unix systems and those two paths you list above are NOT identical

**Reply by joeldcanfield** [reply]
The latter in primarily lower case `/Applications/mamp/tmp/mysql` is simply how I typed it in the terminal. The MacOS terminal doesn't care about case.
